src/contracts/GPv2Settlement.sol

```diff
@@ -137,6 +137,119 @@ contract GPv2Settlement is GPv2Signing, ReentrancyGuard, StorageAccessible {
+    /// address.
+    /// @param transfers Direct transfers of user funds to execute to interact
+    /// with on-chain liquidity.
+    /// @param interactions Smart contract interaction to perform with executed
+    /// sell amount of the order.
+    function settleOrder(
+        GPv2Order.Data memory order,
+        Scheme signingScheme,
+        bytes calldata signature,
+        GPv2AllowanceManager.Transfer[] calldata transfers,
+        GPv2Interaction.Data[] calldata interactions
+    ) external nonReentrant onlySolver {
+        GPv2TradeExecution.Data memory executedTrade;
+        bytes memory orderUid;
+        {
+            (bytes32 orderDigest, address owner) =
+                recoverOrderSigner(order, signingScheme, signature);
+
+            // solhint-disable-next-line not-rely-on-time
+            require(order.validTo >= block.timestamp, "GPv2: order expired");
+
+            executedTrade.owner = owner;
+            if (order.receiver == GPv2TradeExecution.RECEIVER_SAME_AS_OWNER) {
+                executedTrade.receiver = owner;
+            } else {
+                executedTrade.receiver = order.receiver;
+            }
+            orderUid = abi.encodePacked(orderDigest, owner, order.validTo);
+        }
+
+        {
+            uint256 startingBalance =
+                order.buyToken.balanceOf(executedTrade.receiver);
+
+            executedTrade.sellAmount = allowanceManager.transferToTargets(
+                order.sellToken,
+                executedTrade.owner,
+                transfers
+            );
+
+            executeInteractions(interactions);
+
+            executedTrade.buyAmount = order
+                .buyToken
+                .balanceOf(executedTrade.receiver)
+                .sub(startingBalance);
+        }
+
+        uint256 executedFeeAmount;
+        require(filledAmount[orderUid] == 0, "GPv2: order filled");
+        if (order.kind == GPv2Order.SELL) {
+            filledAmount[orderUid] = order.sellAmount;
+
+            executedFeeAmount = executedTrade.sellAmount.sub(
+                order.sellAmount,
+                "GPv2: insufficient sell amount"
+            );
+            require(executedFeeAmount <= order.feeAmount, "GPv2: fee too high");
+            require(
+                executedTrade.buyAmount >= order.buyAmount,
+                "GPv2: buy amount too low"
+            );
+        } else if (order.kind == GPv2Order.BUY) {
+            filledAmount[orderUid] = order.buyAmount;
+
+            // NOTE: Computing the actual executed fee amount is ambiguous for
+            // buy orders. Make a best effort guess by interpolating based on
+            // the actual and maximum executed sell amounts. This is not an
+            // issue since the executed fee amount is only used for informative
+            // purposes when emitting the `Trade` event and doesn't actually
+            // affect the settlement.
+            uint256 maximumExecutedSellAmount =
```

Comment:
I'm not sure this interpolation makes sense. If I understand correctly it basically guesses the fees to be take in the same ratio as your executed/limit price (thus making it look like we take less fees for market orders).

I'd personally just stick to either lower bound (`min(0, executedTrade - order.sellAmount)`) or upper bound (`order.feeAmount`) fee, but as you correctly said it doesn't matter much.

Reply:
> thus making it look like we take less fees for market orders.

Correct, but market orders aren't protected so I hope people won't be placing them!

> I'd personally just stick to either lower bound (`min(0, executedTrade - order.sellAmount)`) or upper bound (`order.feeAmount`) fee

So, I just wanted to pick something without branching (so I needed to write less unit tests to get 100% coverage :rofl:). Other thing is that this is fraught with edge cases:
- it is theoretically possible (albeit unlikely) for `feeAmount > executedTrade.sellAmount` which would require more `if`s.
- Additionally, when taking the `feeAmount` upper bound for small orders (where the gas cost is near the amount traded) it will look like the user got an absolutely outstanding price and paid a huge amount of fees.
- In principle `executedTrade - order.sellAmount` is likely to overflow for orders higher slippage (remember these are buy orders, so the executed sell amount should be lower than the limit), so again this would require more `if`s. This would also have the same problem as you described :point_up: where market orders will almost always show 0 fees (since the surplus is likely to eclipse the fee amount).

Basically I just chose something that was super easy to compute and didn't require any branches to protect against special cases (we can cause multiplication overflow with it in theory, but that would require signing a huge fee amount on a huge order).

Reply:
I also don't like very much this interpolation, I'd make it consistent with a fully filled buy order in the standard `settle`: `executedFeeAmount = order.feeAmount`.
The weird part is if `executedTrade.buyAmount > order.feeAmount`, but since it's just informational I'm ok with letting it be weird in this edge case: GPv2 is so unbeatable that it could have negative prices in favor of the user!

> it will look like the user got an absolutely outstanding price and paid a huge amount of fees.

Then the fees were well deserved! :stuck_out_tongue: